Add info about finding the runtime endpoint (#32132)
kubernetes/website: content/en/docs/tasks/administer-cluster/migrating-from-dockershim/find-out-runtime-you-use.md

node-1       Ready    v1.19.6   containerd://1.4.1
node-2       Ready    v1.19.6   containerd://1.4.1
node-3       Ready    v1.19.6   containerd://1.4.1
```

Find out more information about container runtimes
on [Container Runtimes](/docs/setup/production-environment/container-runtimes/)
page.

## Find out what container runtime endpoint you use {#which-endpoint}

The container runtime talks to the kubelet over a Unix socket using the [CRI
protocol](/docs/concepts/architecture/cri/), which is based on the gRPC
framework. The kubelet acts as a client, and the runtime acts as the server.
In some cases, you might find it useful to know which socket your nodes use. For
example, with the removal of dockershim in Kubernetes 1.24 and later, Docker
Engine cannot communicate with the kubelet if it listens on the
`unix:///var/run/dockershim.sock` endpoint. 

{{<note>}}
If you currently use Docker Engine in your nodes with `cri-dockerd`, you aren't
affected by the dockershim removal.
{{</note>}}

You can check which socket you use by checking the kubelet configuration on your
nodes.

1.  Read the starting commands for the kubelet process:

    ```

>     cat /proc/"$(pgrep kubelet)"/cmdline

Reviewer: only a nit:
```suggestion
    # if you don't have "tr" or "pgrep", you can check the command line more manually
    tr \\0 ' ' < /proc/"$(pgrep kubelet)"/cmdline
```

```
If your runtime shows as Docker Engine, you still might not be affected by the
removal of dockershim in Kubernetes 1.24. [Check the runtime
endpoint](#which-endpoint) to see if you use dockershim. If you don't use
dockershim, you aren't affected. 

For containerd, the output is similar to this:

```none
NAME         STATUS   VERSION   CONTAINER-RUNTIME
node-1       Ready    v1.19.6   containerd://1.4.1
node-2       Ready    v1.19.6   containerd://1.4.1
node-3       Ready    v1.19.6   containerd://1.4.1
```

Find out more information about container runtimes
on [Container Runtimes](/docs/setup/production-environment/container-runtimes/)
page.

## Find out what container runtime endpoint you use {#which-endpoint}

The container runtime talks to the kubelet over a Unix socket using the [CRI
protocol](/docs/concepts/architecture/cri/), which is based on the gRPC
framework. The kubelet acts as a client, and the runtime acts as the server.
In some cases, you might find it useful to know which socket your nodes use. For
example, with the removal of dockershim in Kubernetes 1.24 and later, Docker
Engine cannot communicate with the kubelet if it listens on the

> `unix:///var/run/dockershim.sock` endpoint. 

Reviewer 2: I don't understand this sentence. What communication Between Docker and kubelet is meant here? It is typically kubelet calls into the docker and it uses `docker.sock` to call to it. dokershim.sock is only exposed by kubelet to emulate the CRI from dockershim

{{<note>}}
If you currently use Docker Engine in your nodes with `cri-dockerd`, you aren't
affected by the dockershim removal.
{{</note>}}

You can check which socket you use by checking the kubelet configuration on your
nodes.

1.  Read the starting commands for the kubelet process:

    ```

>     tr \\0 ' ' < /proc/"$(pgrep kubelet)"/cmdline

Reviewer 2: nit: `ps aux | grep [k]ubelet` may be easier

Author: We started with that but decided to go with a more specific command

    ```
    If you don't have `tr` or `pgrep`, check the command line for the kubelet
    process manually.

1.  In the output, look for the `--container-runtime` flag and the

>     `--container-runtime-endpoint` flag.

Reviewer 2: I think just looking at container-runtime should be enough. If not remote than it's dockershim

Author: I made a couple of tweaks.

    *   If your nodes use Kubernetes v1.23 and earlier and these flags aren't
        present, you use the dockershim socket with Docker Engine.
    *   If the `--container-runtime-endpoint` flag is present, check the socket
        name to find out which runtime you use. For example,
        `unix:///run/containerd/containerd.sock` is the containerd endpoint, and

>         `unix:///var/run/dockershim.sock` is the Docker Engine with dockershim

Reviewer 2: `dockershim.sock` is a socket that kubelet exposed, not one that was used to connect to dockershim

Reviewer 2: I don't know what `cri-dockerd` exposes

        endpoint. 

If you use Docker Engine with the dockershim, [migrate to a different runtime](/docs/tasks/administer-cluster/migrating-from-dockershim/change-runtime-containerd/),

> or migrate to [`cri-dockerd`](https://github.com/Mirantis/cri-dockerd) if you

Reviewer 2: why single out the cri-dockerd? it sounds like a recommendation here which is not the intention I think

Reviewer: I think the intention is to enumerate all known CRI-compatible adapters you can use with Docker Engine.

That list is currently: `["cri-dockerd"]`.

Author: I updated the language based on Tim's comment: 

If you use Docker Engine with the dockershim, migrate to a different runtime,
or if you want to continue using Docker Engine in v1.24 and later, migrate to a
CRI-compatible adapter like `cri-dockerd`.

the runtime and its version.

```none
# For docker
NAME         STATUS   VERSION    CONTAINER-RUNTIME
node-1       Ready    v1.16.15   docker://19.3.1
node-2       Ready    v1.16.15   docker://19.3.1
node-3       Ready    v1.16.15   docker://19.3.1
```

> If your runtime is Docker, you still might not be affected by the removal of

Reviewer: ```suggestion
If your runtime is  shows as Docker Engine, you still might not be affected by the removal of
```

```

Find out more information about container runtimes
on [Container Runtimes](/docs/setup/production-environment/container-runtimes/)
page.

## Find out what container runtime endpoint you use {#which-endpoint}

> The container runtime talks to the kubelet uver a Unix socket using the gRPC

Reviewer: ```suggestion
The container runtime talks to the kubelet over a Unix socket using the CRI protocol,
which is based on the gRPC
```
?

framework. The kubelet acts as a client, and the runtime acts as the server.
In some cases, you might find it useful to know which socket your nodes use. For
example, with the removal of dockershim in Kubernetes 1.24 and later, Docker
Engine cannot communicate with the kubelet if it listens on the
`unix:///var/run/dockershim.sock` endpoint. 

{{<note>}}
If you currently use Docker Engine in your nodes with `cri-dockerd`, you aren't
affected by the dockershim removal.
{{</note>}}

You can check which socket you use by checking the kubelet configuration on your
nodes. 

For example, in a cluster created using the `kubeadm` deployment tool, do the
following:

1.  Open `/var/lib/kubelet/kubeadm-flags.env`.
1.  Look for the `--container-runtime-endpoint` flag. If you don't see the flag,
    you're probably using `dockershim`. In this case, [Migrate to another

>     runtime](/tasks/administer-cluster/migrating-from-dockershim/change-runtime-containerd) or [Migrate to cri-dockerd](/tasks/migrate-dockershim-dockerd).

Reviewer: If the kubelet is running, you could look at the command line that it started with. You could also look at things like systemd units.

I'd probably do that even in the presence of a `/var/lib/kubelet/kubeadm-flags.env` because I like to get authoritative info.
